scientific-computing-with-python/learn-special-methods-by-building-a-vector-space/vectors.py

```diff
@@ -0,0 +1,96 @@
+class Vector:
+    space = 'R3'
+
+    def __init__(self, x, y, z):
+        self.x = x
+        self.y = y
+        self.z = z
+        self.norm = (x**2 + y**2 + z**2)**0.5
+
+    def __str__(self):
+        return str((self.x, self.y, self.z))
+    
+    def __repr__(self):
+        return f'Vector({self.x}, {self.y}, {self.z})'
+
+    def __add__(self, other):
+        if not isinstance(other, Vector):
+            return NotImplemented
+        x = self.x + other.x
+        y = self.y + other.y
+        z = self.z + other.z        
+        return Vector(x, y, z)
+    
+    def __sub__(self, other):
+        if not isinstance(other, Vector):
+            return NotImplemented
+        x = self.x - other.x
+        y = self.y - other.y
+        z = self.z - other.z
+        return Vector(x, y, z)
+    
+    def __mul__(self, other):
+        # Scalar multiplication
+        if isinstance(other, (float, int)):
+            x = self.x * other
+            y = self.y * other
+            z = self.z * other            
+            return Vector(x, y, z)
+        
+        # Scalar product
+        elif isinstance(other, Vector):
+            x = self.x * other.x
+            y = self.y * other.y
+            z = self.z * other.z
+            scalar = (x**2 + y**2 + z**2)**0.5
+            return scalar
+        
+    def cross(self, other):
+        # Cross product
+        if not isinstance(other, Vector):
+            raise TypeError('Argument must be a Vector instance')
+        x = self.y * other.z - self.z * other.y
+        y = self.z * other.x - self.x * other.z
+        z = self.x * other.y - self.y * other.x
+        return Vector(x, y, z)
+    
+    def __eq__(self, other):
+        if not isinstance(other, Vector):
+            return NotImplemented
+        if self.x == other.x \
+        and self.y == other.y \
+        and self.z == other.z:
```

Comment:
```suggestion
        if self.x == other.x and self.y == other.y and self.z == other.z:
```
Nitpicking, it's not that long to split it on more lines.